fix: scaling and tuning getting started guide (#29581)
keycloak/keycloak: docs/guides/getting-started/getting-started-scaling-and-tuning.adoc

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/links.adoc" as links>

<@tmpl.guide
title="Scaling"
summary="Get started with {project_name} scaling and tuning">

> Once you have {project_name} started you can next consider adapting your instance to whatever load may be needed. Scaling and tuning can help you:

Reviewer: ```suggestion
After starting {project_name}, consider adapting your instance to the required load using these scaling and tuning guidelines:
```

- minimize resource utilization
- achieve target response times
- minimize database pool contention
- resolve out of memory errors, or excessive garbage collection overhead
- provide higher availability via horizontal scaling

== Vertical Scaling

> As you monitor your {project_name} workload if you see that CPU or memory is either under or over utilized, please consult <@links.ha id="concepts-memory-and-cpu-sizing" /> to better tune the resources available to the Java Virtual Machine (JVM).

Reviewer: ```suggestion
As you monitor your {project_name} workload, check to see the CPU or memory is under or over utilized. Consult <@links.ha id="concepts-memory-and-cpu-sizing" /> to better tune the resources available to the Java Virtual Machine (JVM).
```

> Before increasing the amount of memory available to the JVM, in particular when experiencing an out of memory error, it is best to determine what is contributing to the increased footprint via a heap dump. Excessive response times may also indicate the HTTP work queue is too large and tuning for load shedding would be better than simply providing more memory - see below.

Reviewer: ```suggestion
Before increasing the amount of memory available to the JVM, in particular when experiencing an out of memory error, it is best to determine what is contributing to the increased footprint using a heap dump. Excessive response times may also indicate the HTTP work queue is too large and tuning for load shedding would be better than simply providing more memory. See the following section.
```

=== Common Tuning Options

> {project_name} automatically adjusts the number of used threads based upon how many cores you make available. Manually changing the thread count can improve overall throughput see <@links.ha id="concepts-threads" /> - but this must be done in conjunction with other JVM resources, such as database connections - see <@links.ha id="concepts-database-connections" /> as otherwise you may be moving a bottleneck somewhere else.

Reviewer: ```suggestion
{project_name} automatically adjusts the number of used threads based upon how many cores you make available. Manually changing the thread count can improve overall throughput. For more details,  see <@links.ha id="concepts-threads" />. However, changing the thread count must be done in conjunction with other JVM resources, such as database connections;  otherwise, you may be moving a bottleneck somewhere else. For more details, see <@links.ha id="concepts-database-connections" />.
```

> To limit memory utilization of queued work and to provide for load shedding see <@links.ha id="concepts-threads" anchor="load-shedding" />.

Reviewer: ```suggestion
To limit memory utilization of queued work and to provide for load shedding, see <@links.ha id="concepts-threads" anchor="load-shedding" />.
```

> If you are experiencing timeouts in obtaining database connections, you should consider increasing the number of connections available - see <@links.ha id="concepts-database-connections" />.

Reviewer: ```suggestion
If you are experiencing timeouts in obtaining database connections, you should consider increasing the number of connections available. For more details., see <@links.ha id="concepts-database-connections" />.
```

Author: ```suggestion
If you are experiencing timeouts in obtaining database connections, you should consider increasing the number of connections available. For more details, see <@links.ha id="concepts-database-connections" />.
```

@andymunro there's an extra period in the suggestion, correct?

Reviewer: Yes, the change is tot end the sentence and start a new one with "For more details, ..."

=== Vertical Autoscaling

> Some platforms, such as Kubernetes, provide mechanisms to vertically autoscale. That is not recommended for {project_name} if it requires restarting the server instance - which is currently the case for Java on Kubernetes. You can consider instead providing higher CPU and/or memory limits to allow your JVM to adapt within those limits as needed.

Reviewer: ```suggestion
Some platforms, such as Kubernetes, provide mechanisms to vertically autoscale.  Vertical autoscaling is not recommended for {project_name} if it requires restarting the server instance, which is currently the case for Java on Kubernetes. You can consider instead providing higher CPU and/or memory limits to allow your JVM to adapt within those limits as needed.
```

== Horizontal Scaling

> A single {project_name} instance is susceptible to availability issues. If the instance goes down you will be faced with a full outage until another instance comes up, and unless the '`persistent-user-sessions`' feature is enabled all of your ongoing sessions will be lost. By running at least two cluster members on different machines you will greatly increase the availability of {project_name}.

Reviewer: ```suggestion
A single {project_name} instance is susceptible to availability issues. If the instance goes down, you experience a full outage until another instance comes up. Also, unless the '`persistent-user-sessions`' feature is enabled, all ongoing sessions are lost. By running two orcluster members on different machines, you greatly increase the availability of {project_name}.
```

> There will also be a limit on the amount of concurrent requests a single JVM can effectively handle. Additional server instances can provide roughly linear scaling of throughput until associated resources, such as the database or distributed caching, limit scaling.

Reviewer: ```suggestion
A single JVM has a limit on how many concurrent requests it can handle. Additional server instances can provide roughly linear scaling of throughput until associated resources, such as the database or distributed caching, limit that scaling.
```

> A single {project_name} instance is susceptible to availability issues. If the instance goes down you will be faced with a full outage, and until the persistent sessions feature is available all of your on-going sessions will be lost. By running at least two cluster members on different machines you will greatly increase the availability of {project_name}. 

Reviewer 2: As the persistent sessions feature is available as a preview feature, should we say something like "and unless the `persistent-user-sessions` feature is enabled, all sessions will be lost."

Author: Ok, updated the text.

> In general, it is best to let the {project_name} Operator handle horizontal scaling concerns. When using the Operator simply set the Keycloak custom resource `spec.instances` as desired to horizontally scale - see <@links.ha id="deploy-keycloak-kubernetes" /> for more.

Reviewer: ```suggestion
In general, consider allowing the {project_name} Operator to handle horizontal scaling concerns. When using the Operator, set the Keycloak custom resource `spec.instances` as desired to horizontally scale.  For more details, see <@links.ha id="deploy-keycloak-kubernetes" />.
```

If you are not using the Operator, please review the following:

> * Your instances should be on separate machines if possible to increase availability. On Kubernetes you can use Pod anti-affinitity to enforce this.

Reviewer: ```suggestion
* Higher availability is possible of your instances are on separate machines. On Kubernetes, use Pod anti-affinitity to enforce this.
```

> * You need to use distributed caching and for multi-site clusters external caching for cluster members to share the same session state - see <@links.server id="caching" /> for more on the relevant configuration. The embedded Infinispan cache has horizontal scaling considerations including:

Reviewer: ```suggestion
* Use distributed caching; for multi-site clusters, use external caching for cluster members to share the same session state. For details on the relevant configuration, see <@links.server id="caching" />. The embedded Infinispan cache has horizontal scaling considerations including:
```

> - Your instances will need a way to find each other, see discovery in the server guide.

Reviewer: ```suggestion
- Your instances will need a way to find each other. For more information, see disovery in the https://www.keycloak.org/guides#server[server guides].
```

Reviewer: Note: can you provide a more specific link the Server Guides.  I don't know where discovery is.

Author: @andymunro https://www.keycloak.org/server/caching has more specifics on discovery

Author: ```suggestion
- Your instances need a way to discover each other. For more information, see disovery in the https://www.keycloak.org/server/caching[server guides].
```

Reviewer 3: @shawkins - please use `<@links.server id="caching" />` instead. Also change "disovery" to "discovery". With that change, it is IMHO ready for merging.

> - It is not well suited for clusters spanning multiple availability zones, also known as a stretch clusters. It is recommended that when using the embedded Infinispan cache that the instances all exist in the same availability zone to avoid unnecessary round-trips in the communication that would amplify in the response times. On Kubernetes you can use Pod affinity to enforce this grouping of Pods.

Reviewer: ```suggestion
- This cache is not optimal for clusters that span multiple availability zones, which are also called stretch clusters. When using the embedded Infinispan cache, work to have all instances in one availability zone to avoid unnecessary round-trips in the communication that would amplify in the response times. On Kubernetes you can use Pod affinity to enforce this grouping of Pods.
```

> - It is not well suited for clusters crossing availability zones, also known as a stretch clusters. It is recommended that when using the embedded Infinispan cache that the instances all exist in the same availability zone. On Kubernetes you can use pod affinity to enforce this grouping of pods.

Reviewer 2: `/clusters crossing availability zones/clusters spanning multiple availability zones`?

Author: Added the suggestion.

> - It does not gracefully handle multiple members joining or leaving at the same time - the latter can lead to data loss. On Kubernetes you can use a StatefulSet with the default serial handling to ensure Pods are started and stopped sequentially.

Reviewer: ```suggestion
- This cache does not gracefully handle multiple members joining or leaving concurrently. In particular, members leaving can lead to data loss. On Kubernetes, you can use a StatefulSet with the default serial handling to ensure Pods are started and stopped sequentially.
```

Author: ```suggestion
- This cache does not gracefully handle multiple members joining or leaving concurrently. In particular, members leaving concurrently can lead to data loss. On Kubernetes, you can use a StatefulSet with the default serial handling to ensure Pods are started and stopped sequentially.
```

@andymunro are you good with repeating concurrently here?

Reviewer: It works, but lets change to "at the same time"

> - It does not gracefully handle multiple members joining at the same time. On Kubernetes you can use a StatefulSet using the default of serial handling to ensure Pods are started individually.

Reviewer 2: Multiple members joining simultaneously isn't really an issue, however > 1 member leaving the cluster simultaneously can cause data loss.

Author: @ryanemerson - I'll update this to mention the scale down issue. From #11763, #21108 and other issues, we've also been mentioning a parallel join problem. Is that something later versions of infinispan have addressed? And not related to infinispan, but I believe we've also seen liquibase migration problem when pods start in parallel.

Reviewer 2: > we've also been mentioning a parallel join problem. Is that something later versions of infinispan have addressed?

I'm not aware of an issue that could lead to data loss, or blocking. There maybe some additional network overhead due to rebalances being restarted once all members are added. @pruivo any thoughts?

Reviewer 4: José found some issues. Those are fixed but it may be more out there :) 
* https://issues.redhat.com/browse/ISPN-15432
* https://issues.redhat.com/browse/ISPN-15586
* https://issues.redhat.com/browse/ISPN-15872

Author: The text now mentions both leaving and joining. Does that seem appropriate, or are the join issues sufficiently addressed?

Reviewer 2: I'm happy with the current text :+1:

> To avoid losing service availability when a whole site is not available, see the high availability guide for more information on a multi-site deployment - see <@links.ha id="introduction" />.

Reviewer: ```suggestion
To avoid losing service availability when a whole site is unavailable, see the high availability guide for more information on a multi-site deployment - see <@links.ha id="introduction" />.
```

=== Horizontal Autoscaling

Horizontal autoscaling allows for adding or removing {project_name} instances on demand. Keep in mind that startup times will not be instantaneous and that optimized images should be used to minimize the start time.

> When using the embedded Infinispan cache cluster, dynamically adding or removing cluster members requires Infinispan to perform a rebalancing of the session caches which can get expensive if a lot of entries exist in those caches.

Reviewer: ```suggestion
When using the embedded Infinispan cache cluster, dynamically adding or removing cluster members requires Infinispan to perform a rebalancing of the session caches,m which can get expensive if a many entries exist in those caches.
```

> On Kubernetes the Keycloak custom resource is scalable meaning that it can be targeted by the https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/[built-in autoscaler].

Reviewer: ```suggestion
On Kubernetes, the Keycloak custom resource is scalable meaning that it can be targeted by the https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/[built-in autoscaler].
```

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/links.adoc" as links>

<@tmpl.guide
title="Scaling"
summary="Get started with {project_name} scaling and tuning">

Once you have {project_name} started you can next consider adapting your instance to whatever load may be needed. Scaling and tuning can help you address:

> - minimizing resource utilization

Reviewer 3: My review commit adds a blank line here. In AsciiDoc, those blank lines preceding a list are necessary, as otherwise no list is rendered, but a regular text block.

Author: Thanks for catching that. The default eclipse AsciiDoc preview rendering for some reason is lenient here - I'll switch to a better plugin.

- excessive response times
- database pool contention
- out of memory errors, or excessive garbage collection overhead

> - providing higher availability via horizontal scaling

Reviewer 3: This is mixing positive items (like "minimizing resource utilization") and negative items ("excessive response times"). To be better readable, I'd prefer those items to be all either positive or negative.

Author: Okay, see if the reprasing works for you.

Reviewer 3: Thanks, looks now great to me.

== Vertical Scaling

As you monitor your {project_name} workload if you see that CPU or memory is either under or over utilized, please consult <@links.ha id="concepts-memory-and-cpu-sizing" /> to better tune the resources available to the Java Virtual Machine (JVM).

Before increasing the amount of memory available to the JVM, in particular when experiencing an out of memory error, it is best to determine what is contributing to the increased footprint via a heap dump. Excessive response times may also indicate the HTTP work queue is too large and tuning for load shedding would be better than simply providing more memory - see below.

=== Tuning of the Java Virtual Machine

> Java Virtual Machine Tuning is highly related to the resources available. For example the default garbage collection settings may work well for nominal heap sizes, but may be not ideal for much smaller or larger JVMs. If you are experiencing excessive GC overhead in addition to utilizing the JVM garbage collection tuning options, please open an issue or discussion to raise awareness of your situation.

Reviewer 3: JVM Option tuning is a bit vague, and IMHO it is missing how to set those options. AFAIK it is not possible to set those options in a supported way when using the Operator (only via a Pod Template). 

So it should either be more specific and supported, or removed.

Author: Yes it is unfortunately vague. Given that we've had quite a few issues of late related to this, it seems like we should mention it some way - perhaps just a reference to opening a discussion / issue based upon what symptoms are being experienced.

For the operator you can use an optimized image with environment variables set as needed - however we don't spell out this case anywhere else. Do you want to do that here, or would you rather pare this content down?

Reviewer 3: Looking at the text again today, I think readers will have a hard time to identify a situation of "excessive GC overhead". 

In the KCB project we started to describe the available [JVM metrics](https://www.keycloak.org/keycloak-benchmark/kubernetes-guide/latest/running/jvm/jvm_metrics) and [JVM options](https://www.keycloak.org/keycloak-benchmark/kubernetes-guide/latest/running/jvm/jvm_options). still we haven't come far in describing

* problematic situations (like "excessive GC overhead"),
* how to identify them (by looking at the metrics), and 
* which remedies to take (by specifying JVM options). (cc: @kami619). 

So I would vote to remove or greatly reduce this paragraph, and create a follow-up issue that addresses all three items of the list above. I'd suggest to focus on one situation first (like "excessive GC overhead"), and possibly other situations later (like "out of memory). 

If you hold a different view on this, and consider having the text above is better than not having the text, I am ok to proceed and see what feedback we'll receive on it.

So the bottom line is, you're making the call here, and I'll be happy to approve it. 

BTW, this is the last paragraph we're discussing. All other comments are IMHO resolved.

Author: Ok, removed the JVM level text and opened #31376

Reviewer 3: Thanks, this resolves this one.

== Horizontal Scaling

A single {project_name} instance is susceptible to availability issues. If the instance goes down you will be faced with a full outage until another instance comes up, and unless the '`persistent-user-sessions`' feature is enabled all of your ongoing sessions will be lost. By running at least two cluster members on different machines you will greatly increase the availability of {project_name}.

There will also be a limit on the amount of concurrent requests a single JVM can effectively handle. Additional server instances can provide roughly linear scaling of throughput until associated resources, such as the database or distributed caching, limit scaling.

In general, it is best to let the {project_name} Operator handle horizontal scaling concerns. When using the Operator simply set the Keycloak custom resource `spec.instances` as desired to horizontally scale - see <@links.ha id="deploy-keycloak-kubernetes" /> for more.

If you are not using the Operator, please review the following:

* Your instances should be on separate machines if possible to increase availability. On Kubernetes you can use Pod anti-affinitity to enforce this.

> * You need to use distributed caching and for multi-site clusters external caching for cluster members to share the same session state - see <@links.server id="caching" /> for more on the relevant configuration. The embedded Infinispan cache has horizontal scaling considerations including:

Reviewer 3: So far we haven't advocated for an external Infinispan in single-site setups. They are only recommended for multi-site setups given the additional complexity to set things up. I've updated this paragraph accordingly. 

If we want to recommend an external Infinispan in the future, we would need to have a broader discussion that includes Stian.

Author: Sounds good.

- Your instances will need a way to find each other, see discovery in the server guide.

> - It is not well suited for clusters spanning multiple availability zones, also known as a stretch clusters. It is recommended that when using the embedded Infinispan cache that the instances all exist in the same availability zone to avoid unnecessary round-trips in the communication that would amplify in the response times. On Kubernetes you can use Pod affinity to enforce this grouping of Pods.

Reviewer 3: The previous version didn't say why we wouldn't want stretched clusters. The updated paragraph adds some information. Please review @ryanemerson et al, and expand if needed.
